### PR TITLE
Fix memory leak caused by tracker

### LINKED ISSUE #1586: Ram memory usage problems with track mode
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Other

### Bug

Along with saying hello, I comment that I am tracking with my weights YOLOv8x.pt. The model was trained on a set of construction objects with an input resolution of 1280x1280, I am researching the production of construction machinery via CLI using the command:

I have tried to vary the resolution, the type of model to "m", the frame rate, the track video. The configuration via streaming on youtube works fine, but loading the video from the route is when the use of ram memory increases. I have also tried on A100 80 GB gpu.

The installation module method is pip install -U ultralytics.

The modules installed are these

Package                           Version
--------------------------------- --------------------

- absl-py                           1.4.0
- agate                             1.6.0
- agate-dbf                         0.2.0
- agate-excel                       0.2.3
- agate-sql                         0.5.2
- aiohttp                           3.8.3
- aiosignal                         1.3.1
- anyio                             3.6.2
- argon2-cffi                       21.3.0
- argon2-cffi-bindings              21.2.0
- asttokens                         2.2.1
- astunparse                        1.6.3
- async-timeout                     4.0.2
- attrs                             18.2.0
- awscli                            1.25.91
- Babel                             2.11.0
- backcall                          0.2.0
- beautifulsoup4                    4.11.1
- bleach                            6.0.0
- blis                              0.7.9
- boto3                             1.24.90
- botocore                          1.27.90
- cached-property                   1.5.2
- cachetools                        5.3.0
- catalogue                         2.0.8
- certifi                           2019.11.28
- cffi                              1.15.1
- chardet                           3.0.4
- charset-normalizer                2.1.1
- chex                              0.1.5
- click                             8.1.3
- click-completion                  0.5.2
- click-didyoumean                  0.3.0
- click-help-colors                 0.9.1
- cloudpickle                       2.2.0
- colorama                          0.4.3
- comm                              0.1.2
- confection                        0.0.4
- contourpy                         1.0.7
- csvkit                            1.0.2
- cycler                            0.11.0
- cymem                             2.0.7
- Cython                            0.29.32
- datasets                          2.4.0
- dbfread                           2.0.7
- dbus-python                       1.2.16
- debugpy                           1.6.6
- decorator                         5.1.1
- defusedxml                        0.7.1
- dill                              0.3.5.1
- dm-tree                           0.1.8
- docker-pycreds                    0.4.0
- docutils                          0.16
- entrypoints                       0.4
- et-xmlfile                        1.0.1
- etils                             1.0.0
- exceptiongroup                    1.1.0
- executing                         1.2.0
- fastjsonschema                    2.16.2
- filelock                          3.9.0
- flatbuffers                       1.12
- flax                              0.6.3
- fonttools                         4.38.0
- frozenlist                        1.3.3
- fsspec                            2023.1.0
- future                            0.18.2
- gast                              0.4.0
- gdown                             4.5.1
- gitdb                             4.0.10
- GitPython                         3.1.30
- google-auth                       2.16.0
- google-auth-oauthlib              0.4.6
- google-pasta                      0.2.0
- gql                               3.0.0a6
- gradient                          2.0.6
- gradient-utils                    0.5.0
- graphql-core                      3.1.7
- greenlet                          2.0.1
- grpcio                            1.51.1
- h5py                              3.8.0
- halo                              0.0.31
- huggingface-hub                   0.12.0
- idna                              2.8
- imageio                           2.25.0
- importlib-metadata                6.0.0
- importlib-resources               5.10.2
- iniconfig                         2.0.0
- ipykernel                         6.16.0
- ipython                           8.5.0
- ipython-genutils                  0.2.0
- ipywidgets                        8.0.2
- isodate                           0.6.0
- jax                               0.4.1
- jaxlib                            0.4.1+cuda11.cudnn82
- jdcal                             1.0
- jedi                              0.18.2
- Jinja2                            3.1.2
- jmespath                          1.0.1
- joblib                            1.2.0
- json5                             0.9.11
- jsonify                           0.5
- jsonschema                        4.17.3
- jupyter-client                    7.3.4
- jupyter-contrib-core              0.4.2
- jupyter-contrib-nbextensions      0.7.0
- jupyter_core                      5.1.5
- jupyter-highlight-selected-word   0.2.0
- jupyter-nbextensions-configurator 0.6.1
- jupyter-server                    1.23.5
- jupyter-server-mathjax            0.2.6
- jupyterlab                        3.4.6
- jupyterlab-git                    0.41.0
- jupyterlab-pygments               0.2.2
- jupyterlab_server                 2.19.0
- jupyterlab-snippets               0.4.1
- jupyterlab-widgets                3.0.5
- keras                             2.9.0
- Keras-Preprocessing               1.1.2
- kiwisolver                        1.4.4
- langcodes                         3.3.0
- lap                               0.4.0
- leather                           0.3.3
- libclang                          15.0.6.1
- log-symbols                       0.0.14
- lxml                              4.5.0
- Markdown                          3.4.1
- markdown-it-py                    2.1.0
- MarkupSafe                        2.1.2
- marshmallow                       2.21.0
- matplotlib                        3.6.1
- matplotlib-inline                 0.1.6
- mdurl                             0.1.2
- mistune                           2.0.4
- msgpack                           1.0.4
- multidict                         6.0.4
- multiprocess                      0.70.13
- murmurhash                        1.0.9
- nbclassic                         0.4.8
- nbclient                          0.7.2
- nbconvert                         7.2.9
- nbdime                            3.1.1
- nbformat                          5.7.3
- nest-asyncio                      1.5.6
- networkx                          3.0
- nltk                              3.7
- notebook                          6.5.2
- notebook_shim                     0.2.2
- numpy                             1.23.4
- oauthlib                          3.2.2
- opencv-python                     4.6.0.66
- openpyxl                          3.0.3
- opt-einsum                        3.3.0
- optax                             0.1.4
- orbax                             0.1.0
- packaging                         23.0
- pandas                            1.5.0
- pandocfilters                     1.5.0
- parsedatetime                     2.4
- parso                             0.8.3
- pathtools                         0.1.2
- pathy                             0.10.1
- pexpect                           4.8.0
- pickleshare                       0.7.5
- Pillow                            9.2.0
- pip                               22.3.1
- platformdirs                      2.6.2
- pluggy                            1.0.0
- preshed                           3.0.8
- progressbar2                      4.2.0
- prometheus-client                 0.9.0
- promise                           2.3
- prompt-toolkit                    3.0.36
- protobuf                          3.19.6
- psutil                            5.9.4
- ptyprocess                        0.7.0
- pure-eval                         0.2.2
- pyarrow                           10.0.1
- pyasn1                            0.4.8
- pyasn1-modules                    0.2.8
- pycparser                         2.21
- pydantic                          1.9.2
- Pygments                          2.14.0
- PyGObject                         3.36.0
- pymongo                           3.13.0
- pyparsing                         3.0.9
- pyrsistent                        0.19.3
- PySocks                           1.7.1
- pytest                            7.2.1
- python-apt                        2.0.1
- python-dateutil                   2.8.2
- python-distutils-extra            2.39
- python-slugify                    4.0.0
- python-utils                      3.4.5
- pytimeparse                       1.1.5
- pytz                              2022.7.1
- PyWavelets                        1.4.1
- PyYAML                            5.4.1
- pyzmq                             25.0.0
- regex                             2022.10.31
- requests                          2.28.2
- requests-oauthlib                 1.3.1
- requests-toolbelt                 0.10.1
- requests-unixsocket               0.2.0
- responses                         0.18.0
- rich                              13.2.0
- rsa                               4.7.2
- s3transfer                        0.6.0
- scikit-image                      0.19.3
- scikit-learn                      1.1.2
- scipy                             1.9.2
- seaborn                           0.12.0
- Send2Trash                        1.8.0
- sentence-transformers             2.2.2
- sentencepiece                     0.1.97
- sentry-sdk                        1.14.0
- setproctitle                      1.3.2
- setuptools                        66.1.1
- shellingham                       1.5.0.post1
- shortuuid                         1.0.11
- six                               1.14.0
- smart-open                        6.3.0
- smmap                             5.0.0
- sniffio                           1.3.0
- soupsieve                         2.3.2.post1
- spacy                             3.4.1
- spacy-legacy                      3.0.12
- spacy-loggers                     1.0.4
- spinners                          0.0.24
- SQLAlchemy                        1.4.41
- srsly                             2.4.5
- stack-data                        0.6.2
- tabulate                          0.9.0
- tensorboard                       2.9.1
- tensorboard-data-server           0.6.1
- tensorboard-plugin-wit            1.8.1
- tensorflow                        2.9.2
- tensorflow-estimator              2.9.0
- tensorflow-io-gcs-filesystem      0.30.0
- tensorstore                       0.1.30
- termcolor                         2.2.0
- terminado                         0.17.1
- terminaltables                    3.1.10
- thinc                             8.1.7
- thop                              0.1.1.post2209072238
- threadpoolctl                     3.1.0
- tifffile                          2023.1.23.1
- tinycss2                          1.2.1
- tokenizers                        0.12.1
- tomli                             2.0.1
- toolz                             0.12.0
- torch                             1.12.1+cu116
- torchaudio                        0.12.1+cu116
- torchvision                       0.13.1+cu116
- tornado                           6.1
- tqdm                              4.64.1
- traitlets                         5.8.1
- transformers                      4.21.3
- typer                             0.4.2
- typing_extensions                 4.4.0
- ultralytics                       8.0.55
- Unidecode                         1.1.1
- urllib3                           1.26.14
- wandb                             0.13.4
- wasabi                            0.10.1
- wcwidth                           0.2.6
- webencodings                      0.5.1
- websocket-client                  0.57.0
- Werkzeug                          2.2.2
- wheel                             0.35.1
- widgetsnbextension                4.0.5
- wrapt                             1.14.1
- xgboost                           1.6.2
- xlrd                              1.1.0
- xxhash                            3.2.0
- yarl                              1.8.2
- zipp                              3.11.0

image

### Environment

Python version: Python 3.9.16 (main, Dec  7 2022, 01:11:51) 
OS: Ubuntu 20.04.5 LTS (Focal Fossa)
Specifications: Ultralytics YOLOv8.0.55 🚀 Python-3.9.16 torch-1.12.1+cu116 CUDA:0 (NVIDIA RTX A6000, 48685MiB)
Setup complete ✅ (8 CPUs, 44.1 GB RAM, 131.8/246.0 GB disk)

### Minimal Reproducible Example

yolo track model=/notebooks/runs/detect/train19/weights/best.pt  tracker="botsort.yaml" source="/notebooks/vid4/CraneVistas TowerCrane TimeLapse Diary July 10th 2020 [Ambient Background Whitenoise Music Included.mp4" conf=0.25 iou=0.5 imgsz=1280 save_txt=True save=True save_conf=True half=True

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by pavelbugneac**
Variable `self.remove_stracks` increases without limit which leads to RAM memory depletion over time (especially when running on large videos or rtsp streams). This also leads to slower tracking inference over time since the `sub_stracks` function loops over all items of `self.removed_stracks`. The logic of the fix is to simply keep inmemory the most recently removed tracks instead of the full history. This doesnt have an impact on final results of the tracker.

An alternative would be to simply change line 280 to `self.removed_stracks = removed_stracks` which then only keeps the previous frame history of removed tracks. During own tests, this lead to better reid of lost tracklets, however, a more thorough testing is needed for this alternative approach. For now, the first proposed fix would be the easiest solution

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancement of Byte Tracker algorithm with additional utility function.

### 📊 Key Changes
- Introduced a new static method called `remove_duplicate_stracks` to the Byte Tracker class.

### 🎯 Purpose & Impact
- 🎯 **Purpose**: The new method aims to identify and remove duplicate trackable objects (stracks) by calculating intersection over union (IoU) distances.
- 💥 **Impact**: This will improve the Byte Tracker's ability to track unique objects, reducing errors caused by duplication and potentially enhancing tracking accuracy for users.

### CODE REVIEW

**Comment by pavelbugneac** [comment]
I have read the CLA Document and I sign the CLA

**Comment by pavelbugneac** [comment]
A similar fix was proposed directly to the ByteTrack repo, however, the authors dont seem to be active any longer and arent reviews the latest pull requests. Therefore, submitting it here instead.

**Comment by pavelbugneac** [comment]
Fixes #1586

**Comment by glenn-jocher** [comment]
@pavelbugneac thank you for submitting a fix for issue #1586. It seems that the variable `self.remove_stracks` is leading to RAM memory depletion and slower tracking inference over time. The proposed fix suggests keeping the most recently removed tracks instead of the full history to prevent memory depletion and improve tracking inference. An alternative fix was also suggested, which is to only keep the previous frame history of removed tracks. However, more testing is needed to determine which solution is best. Overall, thank you again for your contribution to Ultralytics YOLOv8.

**Comment by glenn-jocher** [comment]
@pavelbugneac have you profiled reconstructing the list as in this PR with using .remove() on unwanted elements? Not sure which might be faster, i.e.

```python
x = ['a', 'b', 'c']
x.remove('b')
x
Out[10]: ['a', 'c']
```

@Laughing-q what do you think of this? Reasonably we can't extend this array forever, we should limit it to a certain length either very simply with `array = array[:999]` or with some more complicated logic as the PR author has introduced. What's your preferred approach?

**Comment by pavelbugneac** [comment]
> @pavelbugneac have you profiled reconstructing the list as in this PR with using .remove() on unwanted elements? Not sure which might be faster, i.e.
> 
> ```python
> x = ['a', 'b', 'c']
> x.remove('b')
> x
> Out[10]: ['a', 'c']
> ```
> 
> @Laughing-q what do you think of this? Reasonably we can't extend this array forever, we should limit it to a certain length of the most recent tracks, either very simply with `array = array[-999:]` or with some more complicated logic as the PR author has introduced. What's your preferred approach?

@glenn-jocher , I havent tried to use `x.remove()` but I imagine it would also work. The reason for using the list comprehension was to add some logic for the length of the history kept in memory (based on the buffer variable in the config) instead of keeping a static length. Technically simpler slicing would work just as well. So up to you guys.

Maybe worth mentioning, when running about 1 hour inference, tracking slows down by at least 30-40ms per frame (10 fps) due to the for loop in `self.sub_stracks`. This of course can vary depending on amount of objects and other settings in the configs. Also, the second suggestion to set `self.removed_stracks = removed_stracks` seemed to do better at reid but I still need  to test this in case it has other implications.

**Comment by glenn-jocher** [comment]
@pavelbugneac ok got it. If you could quantify some tracking metrics and compare approaches that would be super helpful.

I'll wait on @Laughing-q also as he's one of the principle authors of the tracking implementation so he might have better feedback than myself on this topic.

**Comment by Laughing-q** [comment]
@pavelbugneac @glenn-jocher this is the current logic of bytetracker:

1. prepare `unconfirmed` tracks(new detections with high score, didn't match any tracks) and `confirmed` tracks
2. first match: try to match tracks(both `confirmed` tracks and `lost` tracks) and the `detections with higher conf score`
  a. matched tracks  -> M1
  b. unmatched tracks -> UT1
  c. unmatched detections -> UD1
3. second match: try to match `UT1` and the `detections with lower conf score`
  a. matched tracks  -> M2
  b. unmatched tracks -> UT2 -> mark as `lost`
  c. unmatched detections -> UD2 (UD2 will be ignored in later process)
4. third match: try to match `unconfirmed` tracks and `UD1`
  a. matched tracks -> update the state of tracks from `unconfirmed` to `confirmed`, M3
  b. unmatched tracks -> mark `removed`
  c. unmatched detections -> mark these new detections with high conf score to `unconfirmed` tracks
5. update state -> mark these tracks that `lost` too long as `removed`
6. remove the tracks in `lost` but marked as `removed` in 5th step

I can't see any step that it's necessary to maintain a `self.removed_stracks` to store `removed_stracks` so I think it's okay we just remove `self.removed_stracks` and just use local variable `removed_stracks`.

@pavelbugneac But yes it'd be super helpful if you could quantify some tracking metrics. :)

**Comment by glenn-jocher** [comment]
@Laughing-q it's used in `self.sub_stracks()`, which itself seems to delete old tracks. But what we're really missing is a way to benchmark changes on a tracking dataset.

@pavelbugneac have you measured how long `self.removed_stracks` becomes over time?

Screenshot 2023-04-08 at 16 48 41

**Comment by Laughing-q** [comment]
@glenn-jocher yeah I know, but logcally the most of the tracks in self.removed_tracks is duplicated and outdated, the useful ones are in the local variable removed_tracks.

**Comment by Laughing-q** [comment]
@glenn-jocher still probably it's better we fixed the length of self.removed_tracks i.e 1000 for now. It's the safest way to fix this issue until we work out a benchmark of track.

**Comment by glenn-jocher** [comment]
Ok I'm updating the file and adding a few optimizations in sub_stracks(), i.e.
```
Out[4]: {'a': 0, 'b': 1, 'c': 2}
%timeit 'b' in d
25.6 ns ± 0.031 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
%timeit d.get('b', 0)
46 ns ± 0.0249 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

**Comment by glenn-jocher** [comment]
@pavelbugneac ok I've limited tracks to the last 1000 (though maybe we should reduce this, i.e. to 100?), and tried to optimize sub_stracks() function now. These will be in 8.0.71 soon. Thank you for letting us know about this problem and please check to see if these changes improve performance for long-running tracking tasks!

Screenshot 2023-04-08 at 17 23 14

**Comment by pavelbugneac** [comment]
> @pavelbugneac ok I've limited tracks to the last 1000 (though maybe we should reduce this, i.e. to 100?), and tried to optimize sub_stracks() function now. These will be in 8.0.71 soon. Thank you for letting us know about this problem and please check to see if these changes improve performance for long-running tracking tasks!
> 
> Screenshot 2023-04-08 at 17 23 14

Thanks @glenn-jocher @Laughing-q  for the quick fix! I have been running this now for a few days on my end and it looks good so far. No memory leak as far as I can tell. 
As @Laughing-q mentioned, most tracks in self.removed_tracks are either duplicates or very outdated so reducing to 100 would make sense, although 1000 is also ok for now (it doesnt slow down things too much)

Would you guys still need some metrics on tracking benchmarks? I can try to get some figures in the coming weeks and let you know what I find. I was planning on comparing the tracker outputs in current state vs using the local variable removed_stracks directly (ie without any history).

**Comment by Laughing-q** [comment]
@pavelbugneac yes that would be great! :)

**Comment by alexbarnadas** [comment]
Is this fix available in the main branch? I might be having the same issue

**Comment by glenn-jocher** [comment]
@alexbarnadas yes, the fix has been merged into the main branch. Please pull the latest changes to benefit from the improvements. If you encounter any further issues, don't hesitate to reach out. Happy tracking! 🚀
